### PR TITLE
Poly sqf bug

### PR DESCRIPTION
**Pull request by akhil-dh1**
#### Brief description of what is fixed or changed
Fixed sqf_list function for polynomials. Now, the output of sqf_list will not contain factors of same multiplicity. 
[https://github.com/sympy/sympy/issues/8695#issuecomment-522278244](url)
#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/solvers.py [hunk 1]
```diff
@@ -3679,3 +3679,4 @@ def _norm2(a, b):
 
 from sympy.solvers.bivariate import (
     bivariate_type, _solve_lambert, _filtered_gens)
+
```
Comment (czgdp1807) [inline sympy/solvers/solvers.py]: Add newline.

File: sympy/polys/polytools.py [hunk 2]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
+        numer, denom = together(f).as_numer_denom()
+        numer = Poly(numer, *gens, **args)
+        if denom!=1 and not opt.frac:
+            raise PolynomialError("a polynomial expected, got %s" % f)
+        fp = Poly.sqf_list(numer)
+        coeff = fp[0]
+        fp = [(s.as_expr(), k) for s, k in fp[1]]
+        return coeff ,fp
+    else:
+        raise PolynomialError("a polynomial expected, got %s" % f)
```
Comment (czgdp1807) [inline sympy/polys/polytools.py]: Filters for inputs should be added at the entry point of a function definition.

File: sympy/polys/polytools.py [hunk 3]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
     """
-    return _generic_factor_list(f, gens, args, method='sqf')
-
+    options.allowed_flags(args, ['frac', 'polys'])
+    opt = options.build_options(gens, args)
+    f = sympify(f)
+    if isinstance(f, Expr) and not f.is_Relational:
+        numer, denom = together(f).as_numer_denom()
+        numer = Poly(numer, *gens, **args)
+        if denom!=1 and not opt.frac:
```
Comment (czgdp1807) [inline sympy/polys/polytools.py]: Spaced around `!=`.

File: sympy/polys/polytools.py [hunk 4]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
     (2, [(x + 1, 2), (x + 2, 3)])
 
     """
-    return _generic_factor_list(f, gens, args, method='sqf')
-
+    options.allowed_flags(args, ['frac', 'polys'])
+    opt = options.build_options(gens, args)
+    f = sympify(f)
```
Comment (czgdp1807) [inline sympy/polys/polytools.py]: `_sympify`

**Comment by czgdp1807** [comment]
Unit test should be added for the code added.

**Comment by akhil-dh1** [comment]
Thanks for help, will do as you say.

File: sympy/polys/polytools.py (hunk 2 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: Filters for checking if s is of allowed type to use function _as_expr_?

File: sympy/polys/polytools.py (hunk 4 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: `_sympify` isn't defined in the file

Comment (sylee957) [reply sympy/polys/polytools.py]: `from sympy.core.sympify import _sympify`

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: Thanks. I wanted to confirm that functions with "_" prefix are for calling within i.e they are not accessible outside, is that the case?

File: sympy/polys/polytools.py [hunk 5]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
+    if isinstance(f, Expr) and not f.is_Relational:
```
Comment (jksuom) [inline sympy/polys/polytools.py]: `Relational` is no more a subclass of `Expr` so the condition `f.is_Relational` is not needed.

I would also write

    if not isinstance(f, Expr):
        raise PolynomialError("a polynomial expression expected, got %s" % f)

Then no extra indentation is necessary for the next lines.

File: sympy/polys/polytools.py [hunk 6]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
+        numer, denom = together(f).as_numer_denom()
```
Comment (jksuom) [inline sympy/polys/polytools.py]: This is not needed as there should be no denominator. It there is a denominator, the polynomial creation methods will raise PolynomialError.

File: sympy/polys/polytools.py [hunk 7]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
+        numer = Poly(numer, *gens, **args)
+        if denom!=1 and not opt.frac:
+            raise PolynomialError("a polynomial expected, got %s" % f)
```
Comment (jksuom) [inline sympy/polys/polytools.py]: These two lines can be removed.

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: Okay, thanks your review. Will change these things

File: sympy/polys/polytools.py [hunk 8]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
...
 
     """
-    return _generic_factor_list(f, gens, args, method='sqf')
-
+    options.allowed_flags(args, ['frac', 'polys'])
+    opt = options.build_options(gens, args)
+    f = sympify(f)
+    if isinstance(f, Expr) and not f.is_Relational:
+        numer, denom = together(f).as_numer_denom()
+        numer = Poly(numer, *gens, **args)
```
Comment (jksuom) [inline sympy/polys/polytools.py]: It is probably possible to use `_poly_from_expr(f, opt)`. It may raise an error. Then it should be checked that the result is univariate.

File: sympy/polys/polytools.py (hunk 5 above)
Comment (jksuom) [reply sympy/polys/polytools.py]: This test may not be needed if `_poly_from_expr` is used. It will also sympify the argument.

File: sympy/polys/polytools.py (hunk 6 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: `_poly_from_expr(f, opt)` is allowing denominator to be added to the polynomial. For example:

`sqf_list(  (x**2 + 1)  * (x - 1)**2 * (x - 2)**3 * (x - 3)**3 /(x + 2))`

this generates a polynomial with two generators, `1/(x+2)` and `x`

Comment (jksuom) [reply sympy/polys/polytools.py]: We should check that there is only one generator and that it is a Symbol.

File: sympy/polys/polytools.py [hunk 9]
```diff
@@ -6218,8 +6218,20 @@ def sqf_list(f, *gens, **args):
     (2, [(x + 1, 2), (x + 2, 3)])
 
     """
-    return _generic_factor_list(f, gens, args, method='sqf')
-
+    options.allowed_flags(args, ['frac', 'polys'])
```
Comment (jksuom) [inline sympy/polys/polytools.py]: `_generic_factor_list` can handle also rational fractions if `frac=True`. It may then return factors with negative exponents. That does not apply to `sqf_list` so `'frac'` should be omitted. On the other hand, it might be useful to allow `'all'` that is accepted by the `.sqf_list` method. That should probably be removed from `args` before calling `allowed_flags` and saved (by something like `all = args.pop('all', False)`).

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: okay, but what does adding `all = args.pop('all', False)` actually do?

File: sympy/polys/polytools.py (hunk 8 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: The system hangs when it runs `assert sqf_list(f) == (1, [(x + 1, 40000000000)])` as `_poly_from_expr(f, opt)` is returning the expanded form of polynomial. Is there something to stop this?
Also, what needs to be done for `sqf_list(1)`, should we provide the output or raise an error asking for generator?

**Comment by akhil-dh1** [comment]
The system hangs when it runs assert sqf_list(f) == (1, [(x + 1, 40000000000)]) as _poly_from_expr(f, opt) is returning the expanded form of polynomial. Is there something to stop this?
Also, what needs to be done for sqf_list(1), should we provide the output or raise an error asking for generator?

File: sympy/polys/polytools.py (hunk 9 above)
Comment (jksuom) [reply sympy/polys/polytools.py]: https://docs.python.org/3.7/library/stdtypes.html#dict.pop

**Comment by jksuom** [comment]
I'm not sure that `sqf_list` should necessarily be able to handle artificial examples like `f = (x**2 + 2*x + 1)**20000000000`. However, it would be possible to take care of also such cases by splitting the expression into a product of powers, handling each base polynomial separately, and finally combining the results.

**Comment by akhil-dh1** [comment]
I think then it would be better to improve the original sqf_list method to combine factors with same multiplicities as we have to do the same thing here also(i.e by using Poly's sqf_list method).

**Comment by jksuom** [comment]
One could start by trying to make the expression into a product of powers by `e = together(f)`.
Its factors can be found by `Mul.make_args(e)`, and each `fac` in the tuple is split into a base and exponent: 

    b, e = (fac.base, fac.exp) if fac.is_Pow else (fac, S.One)

The exponent `e` should be positive.

Then each `b` is transformed into a polynomial that should be univariate. If there is a constant `b`, that will go into the coefficient part of the return value. If all polynomials are univariate and have the same generator that is a symbol, then the results from different factors are combined taking their exponents `e` into account.

**Comment by akhil-dh1** [comment]
@jksuom I have now group the factors with same multiplicities. Please review.

**Comment by akhil-dh1** [comment]
Can someone tell me why is the check failing, I haven't even edited the file that's failing.

**Comment by jksuom** [comment]
`_generic_factor_list` is not well suited for `sqf_list` because it does not check the generators.
```
In [1]: e = (x + 1)*(y + 1)

In [2]: sqf_list(e, x)
Out[2]: (y + 1, [(x + 1, 1)])

In [3]: sqf_list(e, y)
Out[3]: (x + 1, [(y + 1, 1)])

In [4]: sqf_list(e)
Out[4]: (1, [(x⋅y + x + y + 1, 1)])
```
Each factor is univariate but their generators are different. Therefore Out[4] is wrong. I think that it is possible to follow the method of `_generic_factor_list` but in a modified form. There is no need to use `as_numer_denom`. If there are denominators, the polynomials will not be univariate, generated by a symbol, and an error can be raised. It should also be checked that there is exactly one generator even if the user has given none. That is easier to do before the Poly objects are turned to expressions by `as_expr`.

**Comment by akhil-dh1** [comment]
@jksuom I don't think _generic_factor_list should be changed as factor_list also uses it. Changes in _generic_factor_list will inflict changes in the output for factor list. I guess then we should first convert to polynomial then do the things or create a separate function for sqf_list. What do you think?

**Comment by jksuom** [comment]
> I don't think _generic_factor_list should be changed as factor_list also uses it.

I don't mean that it should be changed. I think that `sqf_list` should be coded independently as a simplified version of `_generic_factor_list` adding the relevant generator tests.

**Comment by akhil-dh1** [comment]
@jksuom I have changed the code, you can have a look over it.

File: sympy/polys/polytools.py [hunk 10]
```diff
@@ -6218,8 +6216,44 @@ def sqf_list(f, *gens, **args):
...
-
+    options.allowed_flags(args, ['polys'])
+    opt = options.build_options(gens, args)
+    e = together(f)
+    gen_set = set()
+    coeff = 1
+    facs = []
+    for fac in Mul.make_args(e):
+        b, e = (fac.base, fac.exp) if fac.is_Pow else (fac, S.One)
+        _opt = opt.clone(dict(expand=True))
```
Comment (jksuom) [inline sympy/polys/polytools.py]: This clone may not be necessary. It seems that `.expand()` is run anyway unless `opt.expand` is False. https://github.com/sympy/sympy/blob/6cd0b9fade27be198ee24987be0eac709413d297/sympy/polys/polyutils.py#L354-L357

File: sympy/polys/polytools.py [hunk 11]
```diff
@@ -6218,8 +6216,44 @@ def sqf_list(f, *gens, **args):
...
+        if e < 0:
+            raise PolynomialError("a univariate polynomial expected")
+        try:
+            poly = _poly_from_expr(b, _opt)
```
Comment (jksuom) [inline sympy/polys/polytools.py]: Could this be `poly, _ = ...`? The generators are also an attribute of `poly`: `poly.gens`.

File: sympy/polys/polytools.py [hunk 12]
```diff
@@ -6218,8 +6216,44 @@ def sqf_list(f, *gens, **args):
...
+                facs.append((sqf_fac[0] , e*sqf_fac[1]))
+    facs = _sorted_factors(facs, 'sqf')
+    if not opt.polys:
+        facs = [(f.as_expr(), k) for f, k in facs]
+    if facs:
+        new_facs = [facs[0]]
+        for i in range(1, len(facs)):
+            l = len(new_facs)
+            if facs[i-1][1] == facs[i][1]:
+                new_facs[l-1] = (Mul(new_facs[l-1][0], facs[i][0]).expand(), new_facs[l-1][1])
```
Comment (jksuom) [inline sympy/polys/polytools.py]: `Mul` and `expand` are intended for expressions. If the factors are `Poly`s, this may not work as expected:
```
In [1]: Mul(Poly(x), Poly(x + 1))
Out[1]: Poly(x, x, domain='ZZ')⋅Poly(x + 1, x, domain='ZZ')
```

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: @jksuom but I have applied `Mul` only after converting into expression. Nevertheless, what should be used instead of `Mul` and `expand` then? Any suggestions?

Comment (jksuom) [reply sympy/polys/polytools.py]: If `opt.polys=True`, there is no conversion. I would change the order so that conversion would come only after the multiplications which can be the standard polynomial multiplications (with `*`). Those are more efficient than computations with expressions.

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: @jksuom I have made the changes you can see.

File: sympy/polys/polytools.py (hunk 10 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: where can I find the definition of opt.clone() function?

File: sympy/polys/polytools.py [hunk 13]
```diff
@@ -6218,8 +6216,45 @@ def sqf_list(f, *gens, **args):
...
+    coeff = 1
+    facs = []
+    for fac in Mul.make_args(e):
+        b, e = (fac.base, fac.exp) if fac.is_Pow else (fac, S.One)
+        if e < 0:
+            raise PolynomialError("a univariate polynomial expected")
+        try:
+            poly, _ = _poly_from_expr(b, opt)
+        except PolificationFailed:
+            coeff *= b
```
Comment (jksuom) [inline sympy/polys/polytools.py]: Should this have `b**e`?

File: sympy/polys/polytools.py [hunk 14]
```diff
@@ -6218,8 +6216,45 @@ def sqf_list(f, *gens, **args):
...
+        else:
+            gen_set.add(poly.gens)
+            if len(gen_set) > 1:
+                raise PolynomialError("a univariate polynomial expected")
+            cf, sqf_facs = poly.sqf_list()
+            coeff *= cf
```
Comment (jksuom) [inline sympy/polys/polytools.py]: Here as well.

File: sympy/polys/polytools.py (hunk 13 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: Oops! You are right.

Comment (jksuom) [reply sympy/polys/polytools.py]: OK. That is what I expected but was not sure.

**Comment by akhil-dh1** [comment]
@jksuom , @czgdp1807 Is this PR ready to be merged? If so, what should be the steps?

File: sympy/polys/polytools.py [hunk 15]
```diff
@@ -6218,8 +6216,45 @@ def sqf_list(f, *gens, **args):
...
+                new_facs[l-1] = new_facs[l-1][0] * facs[i][0], new_facs[l-1][1]
+            else:
+                new_facs.append(facs[i])
+        if not opt.polys:
+            new_facs = [(f.as_expr(), k) for f, k in new_facs]
+        res = (coeff, new_facs,)
+        return res
+    else:
+        if not opt.polys:
+            facs = [(f.as_expr(), k) for f, k in facs]
```
Comment (jksuom) [inline sympy/polys/polytools.py]: It looks like there are no `facs` at this point.

File: sympy/polys/polytools.py [hunk 16]
```diff
@@ -6218,8 +6216,45 @@ def sqf_list(f, *gens, **args):
...
+    for fac in Mul.make_args(e):
+        b, e = (fac.base, fac.exp) if fac.is_Pow else (fac, S.One)
+        if e < 0:
+            raise PolynomialError("a univariate polynomial expected")
+        try:
+            poly, _ = _poly_from_expr(b, opt)
+        except PolificationFailed:
+            coeff *= b
+        else:
+            gen_set.add(poly.gens)
```
Comment (jksuom) [inline sympy/polys/polytools.py]: `poly.gens` is a tuple. We should add the elements of the tuple.

File: sympy/polys/polytools.py (hunk 15 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: yes, maybe just returning the coeff, facs would be suffice, as the case when facs won't exist is when  the expression is just a constant

File: sympy/polys/polytools.py (hunk 16 above)
Comment (akhil-dh1) [reply sympy/polys/polytools.py]: oh yes, will correct

File: sympy/polys/polytools.py [hunk 17]
```diff
@@ -6231,11 +6231,11 @@ def sqf_list(f, *gens, **args):
         except PolificationFailed:
             coeff *= b
         else:
-            gen_set.add(poly.gens)
+            for gen in poly.gens:
+                gen_set.add(gen)
```
Comment (jksuom) [inline sympy/polys/polytools.py]: How about `gen_set.update(gens)`?

Comment (akhil-dh1) [reply sympy/polys/polytools.py]: yes, update function would also do the thing

**Comment by jksuom** [comment]
I think that the `new_facs` loop could be made more readable by avoiding the use of indices? Maybe something like this:

        new_facs = [facs[0]]
        fac, exp = facs[0]
        for f, e in facs[1:]:
            if exp == e:
                fac *= f
                new_facs[-1] = fac, exp
            else:
                fac, exp = f, e
                new_facs.append((f, e))

Can you find out if this would work?

**Comment by akhil-dh1** [comment]
yah, I think it works. thanks

File: sympy/polys/tests/test_polytools.py [hunk 18]
```diff
@@ -2282,7 +2282,6 @@ def test_sqf():
     raises(ComputationFailed, lambda: sqf_part(4))
 
     assert sqf(1) == 1
-    assert sqf_list(1) == (1, [])
```
Comment (jksuom) [inline sympy/polys/tests/test_polytools.py]: Why was this removed?

File: sympy/polys/tests/test_polytools.py [hunk 19]
```diff
@@ -3263,7 +3262,7 @@ def test_to_rational_coeffs():
 def test_factor_terms():
     # issue 7067
     assert factor_list(x*(x + y)) == (1, [(x, 1), (x + y, 1)])
-    assert sqf_list(x*(x + y)) == (1, [(x, 1), (x + y, 1)])
```
Comment (jksuom) [inline sympy/polys/tests/test_polytools.py]: This could be `raises(PolynomialError, lambda: factor_list(x*(x + y)))`.

This error will not appear if one generator is given: `factor_list(x*(x + y), x)` but then `_sorted_factors` will raise `TypeError`. I think that this should not happen. Maybe something like `sorted(facs, key=lambda f: f[1])` could be used. (It only compares the exponents.)

File: sympy/polys/tests/test_polytools.py [hunk 20]
```diff
@@ -3263,7 +3262,7 @@ def test_to_rational_coeffs():
...
+    assert sqf_list((x**2 + 1)  * (x - 1)**2 * (x - 2)**3 * (x - 3)**3) == (1, [(x**2 + 1, 1), (x - 1, 2), (x**2 -5*x + 6, 3)])
```
Comment (jksuom) [inline sympy/polys/tests/test_polytools.py]:
```suggestion
    assert sqf_list((x**2 + 1) * (x - 1)**2 * (x - 2)**3 * (x - 3)**3) == (1, [(x**2 + 1, 1), (x - 1, 2), (x**2 -5*x + 6, 3)])
```

File: sympy/polys/tests/test_polytools.py (hunk 18 above)
Comment (akhil-dh1) [reply sympy/polys/tests/test_polytools.py]: I wasn't sure if error should be printed or given output should be given out. It's clear now.

File: sympy/polys/tests/test_polytools.py (hunk 20 above)
Comment (akhil-dh1) [reply sympy/polys/tests/test_polytools.py]: Sorry for late response, I was out of town.

File: sympy/polys/tests/test_polytools.py (hunk 19 above)
Comment (akhil-dh1) [reply sympy/polys/tests/test_polytools.py]: Did you mean `raises(PolynomialError, lambda: sqf_list(x*(x + y)))` ?

**Comment by akhil-dh1** [comment]
@jksuom you can check now

File: sympy/polys/tests/test_polytools.py (hunk 19 above)
Comment (jksuom) [reply sympy/polys/tests/test_polytools.py]: Yes, that is what I meant.

**Comment by czgdp1807** [comment]
@akhil-dh1 Any news on this?

**Comment by akhil-dh1** [comment]
> @akhil-dh1 Any news on this?
I think this is ready to be merged. Would be great if someone could review too.

**Comment by akhil-dh1** [comment]
@czgdp1807 @jksuom  I think the main problem of **sqf_list** containing factors of same multiplicity was resolved with issue #8695. I think the issue can be closed.
